doc/tutorials/gapi/face_beautification/face_beautification.markdown

```diff
@@ -0,0 +1,500 @@
+it apriory.
+
+Why isn't memory allocated at the beginning of the `getEyeEllipse()` function?
+The reason is the `cv::ellipse2Poly()` feature: instead of just filling out the
+given array the function assigns to it an array of points created inside;
+therefore, there is no sense to allocate memory before `cv::ellipse2Poly()`
+call.
+
+The rest of the `getEyeEllipse()` function is parameters preparation for the
+`cv::ellipse2Poly()` function call.
+
+![Ellipse2poly illustration](modules/imgproc/doc/pics/ellipse.svg)
+
+What is to be defined:
+- the ellipse center and the X half-axis counted by two eye Points;
+- the Y half-axis counted according to the assumption that an average eye width
+is 1/3 of its length;
+- the start and end angles which are 0 and 180 (see the picture above)
+- the angle delta: how frequently (which causes by how much points) the ellipse
+will be approximated;
+- the inclination angle of axes counted via the
+`getLineInclinationAngleDegrees()` function:
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp toDbl
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp ld_pp_incl
+
+The important thing is the use of the `atan2()` function, which can return
+negative value depending on the `x` and `y` signs; this fact allows to get the
+right angle even in case of upside-down face arrangment (if we put the points in
+the right order, of course).
+
+Back to the `getEyeEllipse()` function, it returns a constructed `Contour`
+to use the RVO.
+
+After the bottom side of an eye is approximated, we just push the eyebrow points
+to the contour in the right order - and the eye is finished!
+
+#### Mouth contour getting {#gapi_fb_ld_mth}
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_old.cpp ld_pp_mth
+
+We have four points of a mouth given: two of them in corners and two others in
+the middle of upper and lower lip. So the mouth can be approximated by two
+patched half-ellipses:
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp ld_pp_mth
+
+At this point Y half-axes are defined without any assumptions, but they are
+different for the upper lip and the lower one.
+
+#### Forehead contour getting {#gapi_fb_ld_fhd}
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_old.cpp ld_pp_fhd
+
+This snippet describes two actions: the approximation of a forehead by
+half-ellipse and pushing the jaw contour to the array in the appropriate order.
+
+The function to approximate a forehead:
+
+@snippet cpp/tutorial_code/gapi/face_beautification/face_beautification_stream.cpp ld_pp_fhd
+
+As we have only jaw points, we are trying to get a half-ellipse based on only
+three points of a jaw: the leftmost, the rightmost and the lowest. Obviously,
+the jaw width is equal to the forehead width and can be counted by the left and
+right points. What about the Y axis, we haven't points to get it directly;
```

Comment:
`Speaking of the Y axis, we have no points ...`, for example

Reply:
OK